chore: Camel K roadmap 2022 (#795)
apache/camel-website: content/blog/2022/03/camel-k-roadmap-2022/index.md

---
title: "Camel K 2022 Roadmap"
date: 2022-03-01
draft: false
authors: [squakez]
categories: ["Roadmap", "Camel K"]

> preview: "A brief overview of what we've planned for 2022"

Reviewer: ```suggestion
preview: "A brief overview of what we've planned for Camel K in 2022"
```

* Merge Camel K runtime into Camel Quarkus
* Multi-architecture images ([#1238](https://github.com/apache/camel-k/issues/1238))
* Modularised build ([#2801](https://github.com/apache/camel-k/issues/2801))
* Nightly builds ([#393](https://github.com/apache/camel-k/issues/393))

We now have a chain of relationships as `Camel K >> Camel K runtime >> Camel Quarkus >> Camel`. Camel K runtime could be seen as a bootstrap of `Camel Quarkus`: this approach simplifies the overall picture and releases dependencies. We may kick off joint initiative with **Camel Quarkus** community to bring the actual **Camel K Runtime**, under the Camel Quarkus umbrella.
Through multi architecture images we will be able to widen our target audience by extending the presence of the **Operator** on different architectures. Nightly build will also let everybody use partial fixes before any official release cycle.

## Installation procedure
We have several installation process options in place. We may need to harmonize them to simplify the maintenance of the installation procedure.

* Document the usage of Kustomize ([#2758](https://github.com/apache/camel-k/issues/2758))

> * Deprecate kamel install in favor of kubectl apply -k / kustomize

Reviewer: ```suggestion
* Deprecate `kamel install` in favor of `kubectl apply -k` / `kustomize`
```

> * Move to descoped OLM model

Reviewer: ```suggestion
* Move to descoped Operator Lifecycle Manager [OLM](https://docs.openshift.com/container-platform/4.9/operators/understanding/olm/olm-understanding-olm.html) model
```

There is some preliminary work introduced in release 1.7 around [Kustomize](https://github.com/apache/camel-k/releases/tag/v1.7.0). We may leverage that in order to identify a replacement for the kamel install. We should analyze how that will happen and the user experience around any new way we’ll propose to the community.
Another thread we should follow is related to the proposal about descoped OLM. This part should be put in a wider context around the OLM project.

## API

> We can start thinking of Camel K as an enablement technology. Then, it makes sense to dedicate some effort in improving the way the technology may be consumed by others. APIs and CRDs are the main scope of this section.

Reviewer: ```suggestion
We can start thinking of Camel K as an enablement technology. Then, it makes sense to dedicate some effort in improving the way the technology may be consumed by others. APIs and Custom Resource Definitions (CRD) are the main scope of this section.
```

* Provide a full description of the APIs

> * Comprehensive CRD structural schemas for tooling

Reviewer: ```suggestion
* Comprehensive Custom Resource Definition (CRD) structural schemas for tooling
```

* Traits configuration schema ([#1614](https://github.com/apache/camel-k/issues/1614))

Implementation of structural schema may be the goal of 2.0 as will probably require many breaking changes. Some draft was realized and it provided discussions we may take in consideration during the feature implementation (see [#2831](https://github.com/apache/camel-k/issues/2831)

## CLI
The CLI is still a very useful tool when interacting with connected environments, in particular in the community. 

> * Fill the gaps for kamel local commands

Reviewer: ```suggestion
* Fill the gaps for `kamel local` commands
```

## Event Driven Architecture
We can think of Camel K as a technology to be used in **Kubernetes** space for event driven architectures. For this reason we will work to enable certain aspects required by EDA.

* Offset management
* Source partitioning and load balancing
* Integration with Keda auto-scalers
* Knative stability
* Kamelets support/stability

Following the strategy of the EDA positioning, we should focus on creating a stable environment for **Knative**, **Kamelets** and **Kafka**. In the same direction we must explore the **Keda** ecosystem and how the first implementation fits into Camel K model.

## Builder
Integration building process needs to be tweaked in order to guarantee scalability. We may revisit parts of the process in order to be more “production grade”.

* Improve concurrency / scalability ([#1784](https://github.com/apache/camel-k/issues/1784))
* Image garbage collection / compaction ([#2736](https://github.com/apache/camel-k/issues/2736))

> * Support OCI registries for universal artifacts management ([#2732](https://github.com/apache/camel-k/issues/2732))

Reviewer: ```suggestion
* Support Open Container Initiative ([OCI](https://opencontainers.org/)) registries for universal artifacts management ([#2732](https://github.com/apache/camel-k/issues/2732))
```

* CI/CD support
* Build Pod compute resources

**[Mvnd](https://github.com/apache/maven-mvnd)** is a good candidate for improving the scalability of the builds. Some preliminary work already exists (see [#2832](https://github.com/apache/camel-k/issues/2832)). Some interesting work is already in place to provide layered images for dependencies (see [#2835](https://github.com/apache/camel-k/issues/2835)). There is certain demand for reducing the quantity of debris produced during the creation of images (ie, garbage collection) and to show how to integrate Camel K with some CI/CD tool.

## Runtime
Other efforts that will let Camel K be more mature are related to features around the runtime.

* Improve route DSL parsing ([#1266](https://github.com/apache/camel-k/issues/1266))

> * Surface Camel health checks into integration status ([#2886](https://github.com/apache/camel-k/issues/2886)

Reviewer: ```suggestion
* Surface Camel health checks into integration status ([#2886](https://github.com/apache/camel-k/issues/2886))
```

* Business vs. technical errors

Route parsing has a great margin of improvement in order to automatically discover components, capabilities, dependencies, etc, and in general to be more reliable. Also, exposing more meaningful ways to query the healthy status of an Integration will let users speed up their troubleshooting (this part is almost completed in the latest releases).

## Observability
Monitoring running Integrations is another field which will let the user feel the maturity of the product.

* Migrate from MicroProfile metrics to Micrometer
* Secured monitoring (see proposal)

> * Grafana dashboards

Reviewer: ```suggestion
* [Grafana](https://grafana.com/) dashboards
```
